GenderPayGap.WebUI/Controllers/Account/ChangePasswordNewController.cs

```diff
@@ -0,0 +1,115 @@
+            IUserRepository userRepository,
+            EmailSendingService emailSendingService)
+        {
+            this.dataRepository = dataRepository;
+            this.userRepository = userRepository;
+            this.emailSendingService = emailSendingService;
+        }
+        
+        // The user asks to change their password
+        // We direct them to a page that asks for their existing and new passwords
+        [HttpGet("change-password-new")]
+        public IActionResult ChangePasswordGet()
+        {
+            ControllerHelper.ThrowIfAdminIsImpersonatingUser(User);
+
+            // Return the Change Personal Details form
+            return View("ChangePasswordNew", new ChangePasswordNewViewModel());
+        }
+
+        [HttpPost("change-password-new")]
+        [ValidateAntiForgeryToken]
+        [PreventDuplicatePost]
+        public async Task<IActionResult> ChangePasswordPost(ChangePasswordNewViewModel viewModel)
+        {
+            // Check all values are provided and NewPassword is at least 8 characters long
+            viewModel.ParseAndValidateParameters(Request, m => m.CurrentPassword);
+            viewModel.ParseAndValidateParameters(Request, m => m.NewPassword);
+            viewModel.ParseAndValidateParameters(Request, m => m.ConfirmNewPassword);
+            
+            // Get the current user
+            User currentUser = ControllerHelper.GetGpgUserFromAspNetUser(User, dataRepository);
+            
+            // Check that passwords are valid
+            ValidatePasswords(viewModel, currentUser);
+
+            if (viewModel.HasAnyErrors())
+            {
+                return View("ChangePasswordNew", viewModel);
+            }
+            
+            userRepository.UpdateUserPasswordUsingPBKDF2(currentUser, viewModel.NewPassword);
+            
+            // send password change notification
+            emailSendingService.SendChangePasswordCompletedEmail(currentUser.EmailAddress);
+
+            // Set up success notification on Manage Account page
+            string nextPageUrl = Url.Action("ManageAccount", "ManageAccount", new {Area = "Account"});
+            StatusMessageHelper.SetStatusMessage(Response, "Your password has been changed successfully", nextPageUrl);
+
+            // Return user to the Manage Account page
+            return LocalRedirect(nextPageUrl);
+        }
+
+        public async void ValidatePasswords(ChangePasswordNewViewModel viewModel, User currentUser)
+        {
+            // Check if current password is correct
+            bool isValidPassword = await userRepository.CheckPasswordAsync(currentUser, viewModel.CurrentPassword);
+            if (!isValidPassword)
+            {
+                viewModel.AddErrorFor(m => m.CurrentPassword, "Could not verify your current password.");
+            }
+
+            // Check if new password has 8 or more characters
+            if (viewModel.NewPassword.Length < 8)

```

Comment:
This caused a NullReferenceException for me because NewPassword wasn't set - we should check that it has a value / it has been parsed successfully before using the value

See how we do this in `AccountCreationController` - e.g.
```csharp
if (viewModel.HasSuccessfullyParsedValueFor(m => m.EmailAddress)
    && viewModel.HasSuccessfullyParsedValueFor(m => m.ConfirmEmailAddress)
    && viewModel.EmailAddress != viewModel.ConfirmEmailAddress)
{
    viewModel.AddErrorFor(
        m => m.ConfirmEmailAddress,
        "The email address and confirmation do not match.");
}
```